diff --git a/lib/src/types/index.d.ts b/lib/src/types/index.d.ts
index cb511409f890179109ce99e39bbdfc3e3e6c1f45..e1aece27e6bd6c3b9fd6c0d1ee5af12490a5c2ca 100644
--- a/lib/src/types/index.d.ts
+++ b/lib/src/types/index.d.ts
@@ -1,14 +1,13 @@
 import React from 'react';
 import { StyleProp, TextProps, TextStyle, TouchableOpacityProps, ViewProps, ViewStyle } from 'react-native';
 export declare type ReactChildren = React.ReactNode;
-export declare type ToastType = 'success' | 'error' | 'info' | (string & {});
 export declare type ToastPosition = 'top' | 'bottom';
-export declare type ToastOptions = {
+export declare type ToastOptions<T extends ToastType = ToastType> = {
     /**
      * Toast type.
      * Default value: `success`
      */
-    type?: ToastType;
+    type?: T;
     /**
      * Style for the header text in the Toast (text1).
      */
@@ -75,18 +74,33 @@ export declare type ToastOptions = {
      * Called on Toast press
      */
     onPress?: () => void;
-    /**
-     * Any custom props passed to the specified Toast type.
-     * Has effect only when there is a custom Toast type (configured via the `config` prop
-     * on the Toast instance) that uses the `props` parameter
-     */
-    props?: any;
-};
+} & (keyof CustomToastParamTypes extends never
+    ? {}
+    : Required<CustomToastParamTypes>[T] extends never
+      ? {}
+      : undefined extends CustomToastParamTypes[T]
+        ? {
+              /**
+               * Any custom props passed to the specified Toast type.
+               * Has effect only when there is a custom Toast type (configured via the `config` prop
+               * on the Toast instance) that uses the `props` parameter
+               */
+              props?: CustomToastParamTypes[T];
+          }
+        : {
+              /**
+               * Any custom props passed to the specified Toast type.
+               * Has effect only when there is a custom Toast type (configured via the `config` prop
+               * on the Toast instance) that uses the `props` parameter
+               */
+              props: CustomToastParamTypes[T];
+          });
+
 export declare type ToastData = {
     text1?: string;
     text2?: string;
 };
-export declare type ToastShowParams = ToastData & ToastOptions;
+export declare type ToastShowParams<T extends ToastType = ToastType> = ToastData & ToastOptions<T>;
 export declare type ToastHideParams = void;
 export declare type BaseToastProps = {
     text1?: string;
@@ -119,9 +133,27 @@ export declare type ToastConfigParams<Props> = {
     onPress: () => void;
     props: Props;
 };
+
+export interface CustomToastParamTypes {}
+
+export interface BuiltinToastParamsTypes {
+    success?: never;
+    error?: never;
+    info?: never;
+}
+
+export declare type ToastType = keyof (BuiltinToastParamsTypes & CustomToastParamTypes);
+
 export declare type ToastConfig = {
-    [key: string]: (params: ToastConfigParams<any>) => React.ReactNode;
-};
+    [key in keyof BuiltinToastParamsTypes]?: (
+        params: ToastConfigParams<BuiltinToastParamsTypes[key]>,
+    ) => React.ReactNode;
+} & {
+    [key in keyof CustomToastParamTypes]-?: (
+        params: ToastConfigParams<CustomToastParamTypes[key]>,
+    ) => React.ReactNode;
+} & Record<string, (params: ToastConfigParams<any>) => React.ReactNode>;
+
 export declare type ToastRef = {
     show: (params: ToastShowParams) => void;
     hide: (params: ToastHideParams) => void;
@@ -130,11 +162,19 @@ export declare type ToastRef = {
  * `props` that can be set on the Toast instance.
  * They act as defaults for all Toasts that are shown.
  */
-export declare type ToastProps = {
-    /**
-     * Layout configuration for custom Toast types
-     */
-    config?: ToastConfig;
+export declare type ToastProps = (keyof CustomToastParamTypes extends never
+    ? {
+          /**
+           * Layout configuration for custom Toast types
+           */
+          config?: ToastConfig;
+      }
+    : {
+          /**
+           * Layout configuration for custom Toast types
+           */
+          config: ToastConfig;
+      }) & {
     /**
      * Toast type.
      * Default value: `success`
